stdpopsim/slim_engine.py

```diff
@@ -648,6 +706,98 @@ def slim_array_string(iterable, indent, width=80):
+
+
+def _enum_dfe_and_intervals(contig):
+    """
+    Returns an iterator which, for every `DFE` in the `contig`, returns a tuple
+    containing a numeric id for said DFE, the DFE object and the list of
+    intervals associated with the DFE. We need to assign each DFE a numeric id
+    both in the slim_makescript and in the _add_dfes_to_metadata, so with this
+    function we can ensure these two steps assigns numeric ids to the DFEs in the
+    same way.
+    """
+    for i, (d, ints) in enumerate(zip(contig.dfe_list, contig.interval_list)):
+        yield i, d, ints
+
+
+def _dfe_to_mtypes(contig):
+    """
+    Assigns a mutation type id to each of the mutation types contained in this
+    `Contig`. This function will return a dictionary with `len(contig.dfe_list)`
+    elements, in which the position of the DFE in `contig.dfe_list` is the key.
+    For each DFE, the dictionary holds a list of tuples that contain the an
+    assigned mutation type id (used in  SLiM) and the `MutationType` object.
+    This is necessary so that we use the same numeric ids to the mutation types
+    in the _add_dfes_to_metadata and slim_makescript steps.
+    """
+    mid = 0
+    dfe_to_mtypes = {}
+    for i, d, _ in _enum_dfe_and_intervals(contig):
+        dfe_to_mtypes[i] = []
+        for mt in d.mutation_types:
+            dfe_to_mtypes[i].append((mid, mt))
+            mid += 1
+    return dfe_to_mtypes
+
+
+def _mutation_types_proportions_slim(d):
+    """
+    Returns a list with the slim proportions for each mutation type in this
+    `stdpopsim.DFE`.  A `MutationType` that is not to be simulated by SLiM will
+    have 0.0 proportion in SLiM, unless this DFE is completely neutral. Because
+    SLiM does not allow a GenomicElementType to have all it's MutationTypes to
+    have 0.0 proportion. This works because all the intervals in this DFE will
+    have 0.0 mutation rate anyway, as computed by
+    `get_msp_and_slim_mutation_rate_maps`.
+    """
+    return [
+        prop if _simulate_mutation_type_in_slim(mt) or d.is_neutral else 0.0
+        for prop, mt in zip(d.proportions, d.mutation_types)
+    ]
+
+
+def _add_dfes_to_metadata(ts, contig):
+    """
+    Adds the DFEs to the top-level metadata of a tree sequence using
+    information in the `contig`.
+    """
+
+    def _get_json(obj):
+        # recursively flattens an object to a dictionary.
+        return json.loads(
+            json.dumps(obj, default=lambda o: getattr(o, "__dict__", str(o)))
+        )
+
+    tables = ts.dump_tables()
+    schema = tables.metadata_schema.asdict()
+    metadata = tables.metadata
+    schema["properties"].update(_raw_stdpopsim_top_level_schema)
+    dfes = _get_json(contig.dfe_list)
+    dfe_to_mtypes = _dfe_to_mtypes(contig)
+    for i, d, ints in _enum_dfe_and_intervals(contig):
+        dfes[i]["intervals"] = ints.tolist()
+        dfes[i]["proportions"] = d.proportions
+        dfes[i]["slim_proportions"] = _mutation_types_proportions_slim(d)
```

Comment:
If we don't put these in metadata then we don't need `_mutation_types_proprotions_slim` to be factored out.